### DISCUSSION TITLE
Optimizing GPU memory utilization

### DISCUSSION THREAD

**Discussion by feddybear** [NVIDIA-NeMo/NeMo#2764] [Q&A] [answered]
Hello NeMo community,

Can somebody point me to a good reading on how to calibrate values for the batch sizes for ASR/TTS training and stuff so I can optimize the memory use of our multi-GPU set ups?

I used to have an experience training with Kaldi where there was an averaging that is computed on the CPU when training with multiple GPUs, making it seem like I was underutilizing our AI training resources.  But now with E2E I want to get more in-depth and understand things better so I can discuss more objectively with those who monitor my GPU activities.

Thanks!

**Comment by titu1994** [comment, marked as answer]
ASR Metrics are calculates on cpu, but almost everything else is on GPU. It should almost not effect utilization though. 

For ASR entire pipeline is on GPU after the initial data loading and pushing onto GPU. For better utilization, use larger models such as Citrinet 1024 or Conformer Large. CTC models have slightly better GPU utilization than RNNT models, but RNNT models converge quite a bit faster so overall wastage is similar. 

Try to find the largest batch size you can fit a few entire epochs without oom, and enable AMP if possible. AMP  may require different hyperparameters though so be careful when tuning hyperparameters

**Reply by feddybear** [reply]
Thanks for the info @titu1994 , now I have a good starting point!  I'm working on Japanese and Tagalog (Philippine) ASR right now.  Which recipe should be a good starting point for me on NeMo?  Currently, I'm only doing the tutorial (number 10 tutorial under ASR) involving QuartzNet 15x5, but would like to upgrade now to the SoTA models ASAP.

**Comment by titu1994** [comment]
You can try out the CTC Finetuning notebook in Nemo 

Notebook for Subword models
https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/ASR_with_Subword_Tokenization.ipynb

Notebook for CTC Finetuning - https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb

For faster convergence you should use Conformer Large models, but for initial runs you can use either Citrinet 512 or Conformer Medium
